artemis-core-client/src/main/java/org/apache/activemq/artemis/core/protocol/core/impl/ActiveMQSessionContext.java

```diff
@@ -189,16 +185,24 @@ public void kill() {
       this.killed = true;
    }
 
+   private void setHandlers() {
+      sessionChannel.setCommandConfirmationHandler(commandConfirmationHandler);
```

Comment:
This was meant to be either newer response handler or the older command confirm handler. The intent was both shouldn't be set. This may be the reason of your dupes issue, what was this needed still even with the async response change?

Reply:
@michaelandrepearce is this ready to be merged in your opinion?

Reply:
I have concern why the handler is getting invoked twice. It shouldnt be possible if its always invoked via the response cache.

Interestingly if i take just the changes justin made in the large message method, then the original failing tests pass. So the other changes seem to be not needed but introduce either issue causing the dupe issue that justin encounters and needs the atomic. I think it needs more discussion and development still. The answer is probably somewhere inbetween

Reply:
I realize that your original intent was that either the confirmation or the response handler would be set here rather than both, and this is definitely the reason for the duplicate confirmation.  However, I couldn't determine any other way to maintain compatibility with existing bridge behavior without setting both.  If this method is changed back to your original semantic (i.e. only set one) then lots of tests in org.apache.activemq.artemis.tests.integration.cluster.bridge.BridgeTest fail.  org.apache.activemq.artemis.tests.integration.client.SendAckFailTest was broken at some point as well.  I will certainly admit that my solution here isn't the most elegant, but I haven't found a more clever way of dealing with the failing tests.  I think eventually the whole confirmation implementation should be refactored to deal with these issues, but I don't see how that's feasible in a minor release.

Reply:
So issue then will be if you dont handle via the cache, the response handlers will keep reference to it, and would slowley leak memory

Reply:
I'm not sure I follow you.  Write a test to demonstrate?

Reply:
see https://github.com/apache/activemq-artemis/pull/2217
hopefully this quick test knocked up explains or demo's the issue.

Reply:
@jbertram
So i think to solve the issue of the responseCache still having reference and the memory leak, we can just add in the code that acks the commands upto the latest, we can just also call the cache, it will mean a double invocation but as you noted for the interim with your flag it actually wont have effect.

Thought? Ill send a pr in a bit to your branch

Reply:
https://github.com/jbertram/activemq-artemis/pull/6